### ISSUE TITLE
Distillation code loss functions

### ISSUE THREAD

**Issue by snaik2016** [huggingface/transformers#2787]
# ❓ Questions & Help
Why compute cross entropy loss from the hard labels in distillation code?

if self.alpha_clm > 0.0:
            shift_logits = s_logits[..., :-1, :].contiguous()
            shift_labels = lm_labels[..., 1:].contiguous()
            loss_clm = self.lm_loss_fct(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
            loss += self.alpha_clm * loss_clm

The model outputs loss when passed with the labels.

## Details

**A link to original question on Stack Overflow**:

**Comment by VictorSanh** [comment]
Hello @snaik2016,
The part of code you're referring to is not a distillation loss. It's the "classic" causal language modeling loss.
Victor

**Comment by snaik2016** [comment]
Not referring to the "distillation loss" just the part of the code where loss is computed in distillation code. The exact same quantity is return by model output when labels are passed.

**Comment by VictorSanh** [comment]
Oh yes, you are right, this could be factorized in.
Just note that you have to be careful with the `ignore_index` and make sure it's coherent with your processing (if I remember correctly, at one point, not all the models were using the same `ignore_index` in the loss computation).
